### PR TITLE
Fix "Incorrect behaviour when using a table constructor expression with any"

### LINKED ISSUE #35436: Incorrect behaviour when using a table constructor expression with `any`
**Description:**
$title.

When `any` is the contextually-expected type, the applicable table type is `table<map<any|error>>`. But the current behaviour seems to consider `any` as a scenario where there is no contextually-expected type.

**Steps to reproduce:**
```ballerina
public function main() {
    // any t1 = table []; // error: cannot infer the member type from table constructor; no values are provided in table constructor
    
    any t2 = table [{a: 1}];
    table<map<any|error>> t3 = <table<map<any|error>>> t2;

    map<any|error> m = {a: error("")};
    t3.add(m); // panic: value type 'map<(any|error)>' inconsistent with the inherent table type 'table<record {| int a; |}>'
}
```

**Affected Versions:**
2201.0.1

### PR DESCRIPTION
**Pull request by gimantha**
Fixes #35436

## Check List 
- [ ] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: tests/jballerina-unit-test/src/test/resources/test-src/types/table/table_value_assign_test.bal [hunk 1]
```diff
@@ -144,12 +149,38 @@ function testAssignKeylessTableValueToTableType() {
...
+    any tbl = table[];
+    table<map<any|error>> a1 = <table<map<any|error>>> tbl;
+    a1.add({"any": "any"});
+    a1.add({"error": error("")});
+
+    any t2 = table [{a: 1}];
+    table<map<any|error>> t3 = <table<map<any|error>>> t2;
+    map<any|error> m = {a: error("")};
+    t3.add(m);
+
```
Comment (lasinicl) [inline tests/jballerina-unit-test/src/test/resources/test-src/types/table/table_value_assign_test.bal:166]: Shall we add a few tests with `any` with table-constructor-expr having a key-specifier too?

**Review by KRVPerera — COMMENTED** [review]
The original example should also work with `var` right.
```Ballerina
public function main() {
    var t2 = table [{a: 1}];
    table<map<any|error>> t3 = <table<map<any|error>>>t2;

    map<any|error> m = {a: error("")};
    t3.add(m); // panic: value type 'map<(any|error)>' inconsistent with the inherent table type 'table<record {| int a; |}>'
}
```
No.. right? Contextually expected type is missing here. So we have to infer from RHS. In that case, it is `table<record{|int a|}>`

**Comment by KRVPerera** [comment]
Yep, sorry for the noise.

**Review by KRVPerera — DISMISSED** [review]

File: tests/jballerina-unit-test/src/test/resources/test-src/types/table/table_value_assign_test.bal (hunk 1 above)
Comment (gimantha) [reply tests/jballerina-unit-test/src/test/resources/test-src/types/table/table_value_assign_test.bal:168]: added tests, according to https://github.com/ballerina-platform/ballerina-spec/issues/1092 it has to be a compile error. So added negative tests

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/types/table/TableNegativeTest.java [hunk 2]
```diff
@@ -187,4 +187,18 @@ public void testTableKeyViolations() {
         validateError(compileResult, index, "duplicate key found in table row key('id') : 'ID2'",
                 136, 9);
     }
+
+    @Test
+    public void testAnyTypedTableWithKeySpecifiers() {
+        CompileResult compileResult = BCompileUtil.compile("test-src/types/table/table-value-any-negative.bal");
+        int index = 0;
+
```
Comment (gimantha) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/types/table/TableNegativeTest.java:195]:
```suggestion
```

File: tests/jballerina-unit-test/src/test/resources/test-src/types/table/table-value-any-negative.bal [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+    ];
+}
+
+function testAssignKeyedTableValueToAnyNegative2() {
+    foo(table key(name) [{name: "Amy", id: 1234}, {name: "John", id: 4567}]);
+}
+
+function foo(any v) {
+
+}
```
Comment (gimantha) [inline tests/jballerina-unit-test/src/test/resources/test-src/types/table/table-value-any-negative.bal:30]:
```suggestion
}

```

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/types/table/TableNegativeTest.java [hunk 4]
```diff
@@ -187,4 +187,17 @@ public void testTableKeyViolations() {
         validateError(compileResult, index, "duplicate key found in table row key('id') : 'ID2'",
                 136, 9);
     }
+
+    @Test
+    public void testAnyTypedTableWithKeySpecifiers() {
+        CompileResult compileResult = BCompileUtil.compile("test-src/types/table/table-value-any-negative.bal");
+        int index = 0;
+        validateError(compileResult, index++,
+                "field name 'name' used in key specifier is not found in table constraint type 'map<(any|error)>'",
```
Comment (pcnfernando) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/types/table/TableNegativeTest.java:196]: Shall we introduce a new error may be which addresses the info gap between how CEP `any` made the table constraint to be map<any|error>?

**Review by KRVPerera — DISMISSED** [review]
LGTM

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/types/table/TableNegativeTest.java (hunk 4 above)
Comment (gimantha) [reply tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/types/table/TableNegativeTest.java:196]: `Key specifier not allowed when contextually expected type is any` ?

Comment (pcnfernando) [reply tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/types/table/TableNegativeTest.java:196]: maybe `Key specifier not allowed when the target type is any`

**Review by pcnfernando — APPROVED** [review]

**Review by lasinicl — APPROVED** [review]

**Pull request merged**